playback/src/audio_backend/rodio.rs

```diff
@@ -0,0 +1,96 @@
+use super::{Open, Sink};
+extern crate rodio;
+use std::{io, thread, time};
+use std::process::exit;
+
+pub struct RodioSink {
+    rodio_sink: rodio::Sink,
+}
+
+fn list_outputs() {
+    println!("Default Audio Device:\n  {:?}", rodio::default_output_device().map(|e| e.name()));
+
+    println!("Available Audio Devices:");
+    for device in rodio::output_devices() {
+        println!("- {}", device.name());
+        // Output formats
+        if let Ok(fmt) = device.default_output_format() {
+            println!("  Default format:\n    {:?}", fmt);
+        }
+        let mut output_formats = match device.supported_output_formats() {
+            Ok(f) => f.peekable(),
+            Err(e) => {
+                println!("Error: {:?}", e);
+                continue;
+            },
+        };
+        if output_formats.peek().is_some() {
+            println!("  All formats:");
+            for format in output_formats {
+                println!("    {:?}", format);
+            }
+        }
+    }
+}
+
+impl Open for RodioSink {
+    fn open(device: Option<String>) -> RodioSink {
+        info!("Using rodio sink");
+
+        let mut rodio_device = rodio::default_output_device().expect("no output device available");
+        if device.is_some() {
+            let device_name = device.unwrap();
+
+            if device_name == "?".to_string() {
+                list_outputs();
+                exit(0)
+            }
+            let mut found = false;
+            for d in rodio::output_devices() {
+                if d.name() == device_name {
+                    rodio_device = d;
+                    found = true;
+                    break;
+                }
+            }
+            if !found {
+                println!("No output sink matching '{}' found.", device_name);
+                exit(0)
+            }
+        }
+        let sink = rodio::Sink::new(&rodio_device);
+
+        RodioSink {
+            rodio_sink: sink,
+        }
+    }
+}
+
+impl Sink for RodioSink {
+    fn start(&mut self) -> io::Result<()> {
+        // More similar to an "unpause" than "play". Doesn't undo "stop".
+        // self.rodio_sink.play();
+        Ok(())
+    }
+
+    fn stop(&mut self) -> io::Result<()> {
+        // This will immediately stop playback, but the sink is then unusable.
```

Comment:
I haven't looked too closely at rodio, but can we not use the stop method to release the sink, and the start method to capture it again? Or is it a case of once closed, it's closed until librespot is reinitialised?

Reply:
Once a sink is closed it seems to set a flag in rodio that means it can't be used anymore. It's also currently the only API to immediately stop playback. The sink can be re-initialized from the rodio device... which seems to work quite well actually. It makes pausing/skipping feel much more responsive. However when using `.stop()` there is consistently a fuzz. I can't figure out any implementation of `fn stop` that don't cause the fuzz. Even setting the volume to 0 makes the sink go fuzzy (regardless of if it's stopped).

I suspect some digging into Rodio/CPAL is needed to get this to work. I may try to find out if the fuzz is windows specific or not tomorrow.

For reference:

```
    fn start(&mut self) -> io::Result<()> {
        // Really an "unpause". Doesn't undo "stop".
        // self.rodio_sink.play();

        // Will cause the existing sink to get finalized (and stop it if it
        // hasn't already... thereby making a fuzz sound).
        // self.rodio_sink = rodio::Sink::new(&self.rodio_device);
        Ok(())
    }

    fn stop(&mut self) -> io::Result<()> {
        // This will immediately stop playback, but the sink is then unusable.
        // It also makes a short fuzz noise when used.
        // self.rodio_sink.stop();

        // Just makes the sink fuzz.
        // self.rodio_sink.set_volume(0.0);

        // This will still fuzz when it gets finalized (be being replaced in play()).
        // self.rodio_sink.pause();

        // We just have to let the current buffer play till the end.
        Ok(())
    }```

Reply:
What do you mean when you say fuzzy? As in it clicks when the sink is opened/closed, or switching back to librespot causes audio distortion?

Reply:
also, I get a compile error if using that line with `rodio_device` from the comment above:

```
error[E0609]: no field `rodio_device` on type `&mut audio_backend::rodio::RodioSink`
  --> playback/src/audio_backend/rodio.rs:79:50
   |
79 |         self.rodio_sink = rodio::Sink::new(&self.rodio_device);
```

Reply:
Here's a windows buildable branch that reproduces the fuzz problem for me on both Ubuntu and Windows if you're curious.

I think I'll have to dig into Rodio and try some smaller examples to hunt this down. Setting the volume seems completely broken too. It would appear [Sinks are not the ideal way to interact with Rodio](https://github.com/tomaka/rodio/issues/171#issuecomment-428978474) which suprises me.

I could have a go at implementing using some lower level APIs in Rodio instead I guess... but at least the implementation in this PR functions ok as far as I know.

Reply:
Can't see the branch? Will give it a go once it's linked here, but if it's a minor issue, would it be worth fixing it later in favour of getting this PR merged sooner rather than later? There's clearly more to be done on this w/ regards to stop method, future audio backend structure, etc.

Reply:
Sorry, here it is: https://github.com/willstott101/librespot/tree/rodio-fuzz

Reply:
building now. is the fuzz consistent, or just start/end of track?

Reply:
I'm not hearing any fuzz when building on OS X with that

Reply:
Ah interesting, yeah it's whenever the stop method is called for me. So pausing does it.

Reply:
Have you tried on a different PC?

Reply:
Yeah I've tried on several PCs on both windows and ubuntu. I'll test on a macbook at some point. I've been working on a CPAL backend using the `sample` crate instead of using Rodio for resampling. I've managed to get that working but it's much more code. Hopefully I can use that as a learning tool to find the fuzz in upstream Rodio though.

Reply:
> CPAL backend using the `sample` crate

would this be an option for me to try to get a librespot with jack + sample rate conversion?

Reply:
The audio code base has changed a fair bit since 2019. Let’s let this PR be.